### ISSUE TITLE
Metric are not logged for Manual Optimization

### ISSUE THREAD

**Issue by awadell1** [PyTorchLightning/pytorch-lightning#7206]
## 🐛 Bug

When Manual optimaztion is enabled (`automatic_optimization = False`), `self.log("loss", loss)` does not get logged.

Or at least does not get logged to the progress bar for: `self.log("loss", loss, prog_bar=True)`. In my use case it also doesn't get logged to wandb, but I didn't replicate that issue here.

### To Reproduce

Use following [**BoringModel**](https://colab.research.google.com/drive/1nzhu4WP0RBY0-55mAPn_vJBq0iWV16K1?usp=sharing)

1) Run the model, notice that loss doesn't appear in the progress bar
2) Change set `self.automatic_optimization = True` in `BoringModel.__init__`
3) Rerun the model, this type the loss does appear, as expected

### Expected behavior

Metrics logged with `self.log` get logged when `self.automatic_optimization = False`

### Environment
```
* CUDA:
	- GPU:
		- Tesla T4
	- available:         True
	- version:           10.1
* Packages:
	- numpy:             1.19.5
	- pyTorch_debug:     False
	- pyTorch_version:   1.8.1+cu101
	- pytorch-lightning: 1.2.10
	- tqdm:              4.41.1
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- 
	- processor:         x86_64
	- python:            3.7.10
	- version:           #1 SMP Thu Jul 23 08:00:38 PDT 2020

### Additional context

The bug also prevent logging to wandb, but not replicated here

**Comment by timothybrooks** [comment]
I have experienced this same issue, although for me it only occurs when using 16bit precision: https://github.com/PyTorchLightning/pytorch-lightning/issues/6833

**Comment by tchaton** [comment]
Hey @timothybrooks, @awadell1 

Would you mind sharing a reproducible script ?

I tried this and it seems to work just fine.

```
def test_multiple_optimizers_logging(tmpdir):
    """
    Tests that only training_step can be used
    """

    class TestModel(BoringModel):

        def __init__(self):
            super().__init__()
            self.automatic_optimization = False

        def training_step(self, batch, batch_idx):
            loss = self.step(batch[0])
            self.log("loss", loss, prog_bar=True)

        def configure_optimizers(self):
            optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
            optimizer_2 = torch.optim.SGD(self.layer.parameters(), lr=0.1)
            return optimizer, optimizer_2

    model = TestModel()
    model.training_epoch_end = None
    model.val_dataloader = None

    limit_train_batches = 2
    trainer = Trainer(
        default_root_dir=tmpdir,
        limit_train_batches=limit_train_batches,
        limit_val_batches=2,
        max_epochs=1,
        log_every_n_steps=1,
        weights_summary=None,
    )

    trainer.fit(model)

    expected = {'epoch', 'loss'}
    logged = set(trainer.logged_metrics.keys())
    assert expected == logged
    expected = {'loss'}
    logged = set(trainer.progress_bar_metrics.keys())
    assert expected == logged
```

<img width="1369" alt="Screenshot 2021-04-27 at 08 51 33" src="https://user-images.githubusercontent.com/12861981/116205531-c54a0580-a735-11eb-8b77-d8b7c4495603.png">

**Comment by timothybrooks** [comment]
Perhaps this will help: https://github.com/PyTorchLightning/pytorch-lightning/issues/6833#issuecomment-814501216

**Comment by timothybrooks** [comment]
Also worth trying 16bit, since that is only when it occurred for me.

**Comment by tchaton** [comment]
Hey @timothybrooks,

Resolve the bug, it was pretty silly. optimiser.step calls `call_hook("on_after_backward")` which was erasing Result object.

Pushing a fix soon.

Best,
T.C

**Comment by awadell1** [comment]
It looks like you guys already got it, so huge thank you!

But if your looking for a test case, see here: https://colab.research.google.com/drive/1nzhu4WP0RBY0-55mAPn_vJBq0iWV16K1?usp=sharing

I've added the assert code from [above](https://github.com/PyTorchLightning/pytorch-lightning/issues/7206#issuecomment-827396972) to my example notebook ( I really should have left the URL unformatted)

Thanks for finding this so quickly,

Best,

Alex
